akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -444,108 +568,173 @@ private[akka] class Shard(
-        stash()
+          log.debug("Start entity [{}] while a write already in progress. Marking as pending", entityId)
+        entities.rememberingStart(entityId, ackTo = Some(sender))
+      } else {
+        // it's already running, ack immediately
+        sender() ! ShardRegion.StartEntityAck(entityId, shardId)
+      }
+
+    // below cases should handle same messages as in idle
+    case Terminated(ref) =>
+      entities.entityId(ref) match {
+        case OptionVal.Some(entityId) =>
+          entities.entityState(entityId) match {
+            case RememberingStop(_) =>
+              log.debug("Stop of [{}] arrived, already is among the pending stops.", entityId)
+            case Active(_) =>
+              // FIXME should trigger backoff + restart?
+              log.warning("Stop of [{}] arrived, should lead to backoff restart, not implemented yet.", entityId)
+            case Passivating(_) =>
+              // will go in next batch update
+              log.debug("Stop of [{}] arrived while updating, adding it to batch of pending stops.", entityId)
+              entities.rememberingStop(entityId, Passivating)
+            case unexpected =>
+              log.warning("Got a terminated for [{}], entityId [{}] which is in state [{}]", ref, entityId, unexpected)
+          }
+        case OptionVal.None =>
+          // FIXME what would this Terminated be? Remember entities store stop?
+          stash()
+      }
+    case _: CoordinatorMessage         => stash()
+    case _: StartEntityInternal        => stash()
+    case _: RememberEntityCommand      => stash()
+    case l: LeaseLost                  => receiveLeaseLost(l)
+    case _: ShardRegion.StartEntityAck =>
+      // FIXME if we got an ack it was started elsewhere we should probably make sure we don't have it in current or pending starts?
+      stash()
+    case ShardRegion.Passivate(stopMessage) =>
+      if (VerboseDebug)
+        log.debug(
+          "Passivation of [{}] arrived while updating.",
+          entities.entityId(sender()).getOrElse(s"Unknown actor ${sender()}"))
+      passivate(sender(), stopMessage)
+    case _: ShardRegionCommand           => stash()
+    case msg: ShardQuery                 => receiveShardQuery(msg)
+    case PassivateIdleTick               => stash()
+    case msg: RememberEntityStoreCrashed => rememberEntityStoreCrashed(msg)
+    case msg if extractEntityId.isDefinedAt(msg) =>
+      deliverMessage(msg, sender())
+    case msg =>
+      // shouldn't be any other message types, but just in case
+      log.warning(
+        "Stashing unexpected message [{}] while waiting for remember entities update of starts [{}], stops [{}]",
+        msg.getClass,
+        update.started.mkString(", "),
+        update.stopped.mkString(", "))
+      stash()
+
+  }
+
+  def onUpdateDone(starts: Set[EntityId], stops: Set[EntityId]): Unit = {
+    // entities can contain both ids from start/stops and pending ones, so we need
```

Comment:
what are the cases where this happens, it gets a message when a different write is happening, and another that stops it? In that case can we just not store it at all?

Or the other way around, we get a stop, there passivated happens, and then we get another message for the entity, we could just not store the stop or the start?

Reply:
The scenarios I think could be ones to cover:
 * when a stop is batched, and a message for it arrives before the batch is handled
 * when a start is batched, and a passivate or entity start ack completes before the batch is handled

When the stop or start is actual write in progress there's nothing to do about it but handle it later, it could be that is a problem with the current design as there is nowhere to track that, if we update the state for the entity we lose the previous - when we get here we will remove the id as stopped/started before we try figuring out what ids are up for next batch.